### PR TITLE
Remove the exposure of SharedDirectory API

### PR DESCRIPTION
**Pull request by clarenceli-msft**
Hide the class implementation of `SharedDirectory` and update the corresponding exports, a follow up fix of https://github.com/microsoft/FluidFramework/pull/19717

[AB#5799](https://dev.azure.com/fluidframework/internal/_workitems/edit/5799)

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/directoryEndToEndTests.spec.ts [hunk 1]
```diff
@@ -1027,7 +1027,7 @@ describeCompat("SharedDirectory orderSequentially", "NoCompat", (getTestObjectPr
 		assert.notEqual(error, undefined, "No error");
 		assert.equal(error?.message, errorMessage, "Unexpected error message");
 		assert.equal(containerRuntime.disposed, false);
-		assert.equal(sharedDir.countSubDirectory(), 0);
+		assert.equal(sharedDir.countSubDirectory?.(), 0);
```
Comment (CraigMacomber) [inline packages/test/test-end-to-end-tests/src/test/directoryEndToEndTests.spec.ts:1030]: I think the only implementation of ISharedDirectory is SharedDirectory, so it seems like this methods should be able to be made non-optional on it, allowing these tests (and possible apps using it as well) to remain unchanined.

Comment (CraigMacomber) [reply packages/test/test-end-to-end-tests/src/test/directoryEndToEndTests.spec.ts:1030]: might be good to try and figure out why that method was made optional in the first place as its doc's don't say. Maybe there are third parties implementing the interface that can't implement that?

Comment (clarenceli-msft) [reply packages/test/test-end-to-end-tests/src/test/directoryEndToEndTests.spec.ts:1030]: I reviewed the update history of this method, and it appears that the FFX repo did implement the `IDirectory` interface. Somehow we made this method optional. https://github.com/microsoft/FluidFramework/pull/9766/files#r844412897
Add @jatgarg here

File: .changeset/clever-cats-jump.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+"fluid-framework": minor
+"@fluidframework/map": minor
+---
+
+DDS classes are no longer publicly exported
+
+ShareDirectory and SharedTree now only export their factories and the interface types.
```
Comment (scarlettjlee) [inline .changeset/clever-cats-jump.md:8]:
```suggestion
SharedDirectory and SharedTree now only export their factories and the interface types.
```

Comment (scarlettjlee) [inline .changeset/clever-cats-jump.md:8]: When did this change happen for SharedTree? Within the same version? We're about to branch RC2.

Comment (clarenceli-msft) [reply .changeset/clever-cats-jump.md]: It is a typo, already fixed

File: packages/dds/map/src/index.ts [hunk 3]
```diff
@@ -98,3 +98,46 @@ export const SharedMap = {
  * This alias is for legacy compat from when the SharedMap class was exported as public.
  */
 export type SharedMap = ISharedMap;
+
+/**
+ * {@inheritDoc ISharedDirectory}
+ * @sealed
+ * @alpha
+ */
+export const SharedDirectory = {
```
Comment (Abe27342) [inline packages/dds/map/src/index.ts:107]: This should go in a file other than `index.ts`. Putting it in the same file the factory is defined in seems reasonable (that also has the side-benefit of cutting the circular dep between the DDS class and its factory, if they're defined in different files)

File: packages/dds/map/src/index.ts [hunk 4]
```diff
@@ -98,3 +98,46 @@ export const SharedMap = {
  * This alias is for legacy compat from when the SharedMap class was exported as public.
  */
 export type SharedMap = ISharedMap;
+
+/**
+ * {@inheritDoc ISharedDirectory}
```
Comment (Abe27342) [inline packages/dds/map/src/index.ts:103]: Inheriting doc seems inappropriate here. Something like this seems better:

`Entrypoint for {@link ISharedDirectory} creation`.

However, I think once Craig's PR here merges: https://github.com/microsoft/FluidFramework/pull/20028, you'll have a clearer way to document using `ISharedObjectKind` (and can just inherit docs from that)

File: packages/test/test-version-utils/api-report/test-version-utils.api.md [hunk 5]
```diff
@@ -79,7 +79,10 @@ export const DataRuntimeApi: {
     dds: {
         SharedCell: typeof cell.SharedCell;
         SharedCounter: typeof counter.SharedCounter;
-        SharedDirectory: typeof map.SharedDirectory;
+        SharedDirectory: {
```
Comment (Abe27342) [inline packages/test/test-version-utils/api-report/test-version-utils.api.md:82]: seems like you should be able to keep previous `typeof` syntax, no?

**Review by Abe27342 — COMMENTED** [review]
looks good, mostly some organizational nits. We'll need to do some merging with concepts added in https://github.com/microsoft/FluidFramework/pull/20028. So keep an eye on that before merging.

File: packages/dds/map/src/index.ts (hunk 3 above)
Comment (clarenceli-msft) [reply packages/dds/map/src/index.ts]: I combined the factory and object within the same file, but it appears inevitable that we have to import and export `SharedMap/SharedDirectory` within the same file. I did some tricks such as `import { SharedMap as SharedMapInternal }`, not sure if  there is a better approach. (in the commit https://github.com/microsoft/FluidFramework/pull/20030/commits/0de8bef448bcb105948307199dec33c5db6988d9#diff-8475d479f31499dd12b9b24699ad7b1cc0972e7eee079ad4e0e0281ae6387d1f)

File: packages/dds/map/src/index.ts (hunk 4 above)
Comment (Abe27342) [reply packages/dds/map/src/index.ts]: Yeah, just referencing `ISharedDirectory` is fine. inheritdoc has a lot of weird quirks in our toolchain IIRC.

File: packages/test/test-version-utils/api-report/test-version-utils.api.md [hunk 6]
```diff
@@ -81,7 +81,7 @@ export const DataRuntimeApi: {
     dds: {
         SharedCell: typeof cell.SharedCell;
         SharedCounter: typeof counter.SharedCounter;
-        SharedDirectory: typeof map.SharedDirectory;
+        SharedDirectory: ISharedObjectKind<map.ISharedDirectory>;
```
Comment (Abe27342) [inline packages/test/test-version-utils/api-report/test-version-utils.api.md:84]: The change here seems unnecessary (the typeof statement evaluates to the same thing). I'd rather change the `SharedMap` one back in test-version-utils. I can do that separately though.

**Review by Abe27342 — COMMENTED** [review]
Can `ISerializableValue` and `ISerializedValue` be removed from the public API as well? Doesn't look like they're referenced anymore.

**Comment by clarenceli-msft** [comment]
> Can `ISerializableValue` and `ISerializedValue` be removed from the public API as well? Doesn't look like they're referenced anymore.

Moved them to `./internalInterface.ts` and removed the exposure of them from API

**Review by Abe27342 — APPROVED** [review]

File: .changeset/clever-cats-jump.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+---
+"fluid-framework": minor
+"@fluidframework/map": minor
```
Comment (alexvy86) [inline .changeset/clever-cats-jump.md:3]: There was a discussion about this on another PR, I'm on the side that we should try to be deliberate about this bit in a changeset. What we're doing here qualifies as a major change.

**Review by alexvy86 — APPROVED** [review]
Devtools change is good. Left a general comment about the changeset though.

File: .changeset/clever-cats-jump.md (hunk 7 above)
Comment (clarenceli-msft) [reply .changeset/clever-cats-jump.md:3]: oops it is automatically merged in, I will create a small PR to update it.

**Pull request merged**